iscservice/demo/iscdemo.c

```diff
@@ -0,0 +1,607 @@
+/* -*- C -*- */
+/*
+ * Copyright (c) 2021 Seagate Technology LLC and/or its Affiliates
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ */
+
+#include <unistd.h>       /* getopt */
+
+#include "lib/memory.h"   /* m0_alloc m0_free */
+#include "rpc/conn.h"     /* m0_rpc_conn_addr */
+#include "rpc/session.h"  /* iop_session */
+#include "xcode/xcode.h"
+#include "conf/schema.h"  /* M0_CST_ISCS */
+#include "layout/layout.h" /* M0_DEFAULT_LAYOUT_ID */
+#include "layout/plan.h"  /* m0_layout_plan_build */
+#include "motr/client.h"
+
+#include "util.h"
+#include "common.h"
+#include "libdemo.h"      /* mm_result */
+#include "libdemo_xc.h"   /* isc_args_xc */
+
+
+enum isc_comp_type {
+	ICT_PING,
+	ICT_MIN,
+	ICT_MAX,
+};
+
+static int op_type_parse(const char *op_name)
+{
+	if (op_name == NULL)
+		return -EINVAL;
+	else if (!strcmp(op_name, "ping"))
+		return ICT_PING;
+	else if (!strcmp(op_name, "min"))
+		return ICT_MIN;
+	else if (!strcmp(op_name, "max"))
+		return ICT_MAX;
+	else
+		return -EINVAL;
+
+}
+
+static int minmax_input_prepare(struct m0_buf *out, struct m0_fid *comp_fid,
+				struct m0_layout_io_plop *iop,
+				uint32_t *reply_len, enum isc_comp_type type)
+{
+	int           rc;
+	struct m0_buf buf = M0_BUF_INIT0;
+	struct isc_targs ta = {};
+
+	if (iop->iop_ext.iv_vec.v_nr == 0) {
+		fprintf(stderr, "at least 1 segment required\n");
+		return -EINVAL;
+	}
+	ta.ist_cob = iop->iop_base.pl_ent;
+	rc = m0_indexvec_mem2wire(&iop->iop_ext, iop->iop_ext.iv_vec.v_nr, 0,
+				  &ta.ist_ioiv);
+	if (rc != 0)
+		return rc;
+	rc = m0_xcode_obj_enc_to_buf(&M0_XCODE_OBJ(isc_targs_xc, &ta),
+				     &buf.b_addr, &buf.b_nob);
+	if (rc != 0)
+		return rc;
+
+	*out = M0_BUF_INIT0; /* to avoid panic */
+	rc = m0_buf_copy_aligned(out, &buf, M0_0VEC_SHIFT);
+	m0_buf_free(&buf);
+
+	if (type == ICT_MIN)
+		isc_fid_get("comp_min", comp_fid);
+	else
+		isc_fid_get("comp_max", comp_fid);
+
+	*reply_len = CBL_DEFAULT_MAX;
+
+	return rc;
+}
+
+static int ping_input_prepare(struct m0_buf *buf, struct m0_fid *comp_fid,
+			      uint32_t *reply_len, enum isc_comp_type type)
+{
+	char *greeting;
+
+	*buf = M0_BUF_INIT0;
+	greeting = m0_strdup("Hello");
+	if (greeting == NULL)
+		return -ENOMEM;
+
+	m0_buf_init(buf, greeting, strlen(greeting));
```

Comment:
This is a little confusing to me.

The min/max is encoding "struct isc_targs" into the output 'buf', and at server side, the same struct can be decoded.

But the ping is encoding a string "hello" into output 'buf'.
Why are they different?
I know for the ping, "struct isc_targs" is not used in comp. But is it better to send the same data to server?
Otherwise, it is confusing.

Also, the encoding of comp parameters can be a common function and be used by all comp.
The difference of different COMPutation is the fid.

Reply:
The computation function can encode anything to the input buffer. The ping "computation" does not need to read the user data from the cob (as min/max does), so there is no need to use isc_targs for it.

Reply:
OK.
`The computation function can encode anything to the input buffer.`
The author is responsible to encode and decode its own input correctly.